### PR TITLE
policy for systemd-fstab-generator

### PR DESCRIPTION
**Pull request by bauen1**
Implementing a policy for systemd-fstab-generator instead of allowing `initrc_t` to write to `init_runtime_t`.

Tested on:
- [x] debian minimal install (enforcing=1)

Comments are appreciated.

### CODE REVIEW

File: policy/modules/system/systemd.te [hunk 1]
```diff
@@ -254,6 +258,36 @@ files_read_etc_files(systemd_binfmt_t)
 
 fs_register_binary_executable_type(systemd_binfmt_t)
 
+#######################################
+#
+# fstab generator local policy
+#
+
+gen_require(`
+    type init_runtime_t;
```
Comment (doverride) [inline policy/modules/system/systemd.te]: Not allowed

File: policy/modules/system/systemd.te [hunk 2]
```diff
@@ -254,6 +258,36 @@ files_read_etc_files(systemd_binfmt_t)
...
+')
+
+# /run/systemd/generator
+manage_dirs_pattern(systemd_fstab_generator_t, init_runtime_t, init_runtime_t)
+allow systemd_fstab_generator_t init_runtime_t:file { create_file_perms write };
+allow systemd_fstab_generator_t init_runtime_t:lnk_file create_lnk_file_perms;
+
+kernel_read_kernel_sysctls(systemd_fstab_generator_t)
+kernel_read_system_state(systemd_fstab_generator_t)
+dev_write_kmsg(systemd_fstab_generator_t)
```
Comment (doverride) [inline policy/modules/system/systemd.te]: use the "log_parse_env" interface

File: policy/modules/system/init.if [hunk 3]
```diff
@@ -1380,6 +1380,61 @@ interface(`init_list_pids',`
...
+## <summary>
+##  Create symbolic links in the /run/systemd directory.
+## </summary>
+## <param name="domain">
+##  <summary>
+##  Domain allowed access.
+##  </summary>
+## </param>
+#
+interface(`init_create_pids_symlinks', `
```
Comment (pebenito) [inline policy/modules/system/init.if]: `init_manage_pid_symlinks`. Similar naming below.

File: policy/modules/system/systemd.te [hunk 4]
```diff
@@ -258,6 +262,31 @@ files_read_etc_files(systemd_binfmt_t)
...
+
+init_search_pids(systemd_fstab_generator_t)
+
+fstools_exec(systemd_fstab_generator_t)
+
+files_search_pids(systemd_fstab_generator_t)
+
+# for /etc/fstab
+files_read_etc_files(systemd_fstab_generator_t)
+
```
Comment (pebenito) [inline policy/modules/system/systemd.te]: Please reorder to alphabetical order.

**Comment by bauen1** [comment]
> Please reduce the number of commits. The one whitespace change should be its own commit.

I'll adjust my commit style for future pull-requests.

**Comment by bauen1** [comment]
Please don't merge, I still need to test this.

**Comment by bauen1** [comment]
This PR has been tested on a debian minimal install.

File: policy/modules/system/init.if [hunk 5]
```diff
@@ -1380,6 +1380,61 @@ interface(`init_list_pids',`
...
+## <summary>
+##  Create and write files in the /run/systemd directory.
+## </summary>
+## <param name="domain">
+##  <summary>
+##  Domain allowed access.
+##  </summary>
+## </param>
+#
+interface(`init_create_write_pid_files', `
```
Comment (pebenito) [inline policy/modules/system/init.if]: Split into init_create_pid_files() and init_write_pid_files()

**Comment by bauen1** [comment]
Tested.

**Comment by dsugar100** [comment]
I'm seeing the following denials on RHEL7

[   19.046773] type=1400 audit(1584424888.832:5): avc:  denied  { use } for  pid=1639 comm="systemd-fstab-g" path="/dev/null" dev="devtmpfs" ino=2051 scontext=system_u:system_r:systemd_fstab_generator_t:s0 tcontext=system_u:system_r:kernel_t:s0 tclass=fd permissive=1
[   19.048632] type=1400 audit(1584424888.834:6): avc:  denied  { dac_override } for  pid=1639 comm="systemd-fstab-g" capability=1  scontext=system_u:system_r:systemd_fstab_generator_t:s0 tcontext=system_u:system_r:systemd_fstab_generator_t:s0 tclass=capability permissive=1
[   19.048644] type=1400 audit(1584424888.834:7): avc:  denied  { write } for  pid=1639 comm="systemd-fstab-g" name="/" dev="sysfs" ino=1 scontext=system_u:system_r:systemd_fstab_generator_t:s0 tcontext=system_u:object_r:sysfs_t:s0 tclass=dir permissive=1
[   19.048694] type=1400 audit(1584424888.834:8): avc:  denied  { write } for  pid=1639 comm="systemd-fstab-g" name="generator" dev="tmpfs" ino=16386 scontext=system_u:system_r:systemd_fstab_generator_t:s0 tcontext=system_u:object_r:systemd_unit_t:s0 tclass=dir permissive=1
[   19.048702] type=1400 audit(1584424888.834:9): avc:  denied  { add_name } for  pid=1639 comm="systemd-fstab-g" name="-.mount" scontext=system_u:system_r:systemd_fstab_generator_t:s0 tcontext=system_u:object_r:systemd_unit_t:s0 tclass=dir permissive=1
[   19.048727] type=1400 audit(1584424888.834:10): avc:  denied  { create } for  pid=1639 comm="systemd-fstab-g" name="-.mount" scontext=system_u:system_r:systemd_fstab_generator_t:s0 tcontext=system_u:object_r:systemd_unit_t:s0 tclass=file permissive=1
[   19.048741] type=1400 audit(1584424888.834:11): avc:  denied  { write open } for  pid=1639 comm="systemd-fstab-g" path="/run/systemd/generator/-.mount" dev="tmpfs" ino=12979 scontext=system_u:system_r:systemd_fstab_generator_t:s0 tcontext=system_u:object_r:systemd_unit_t:s0 tclass=file permissive=1

It looks like the bulk of the errors are being caused by /run/systemd/generator being labeled systemd_unit_t on my system and this commit is expecting that directory to be init_runtime_t.  Any idea what is creating the directory /run/systemd/generator?

**Comment by doverride** [comment]
Dave Sugar <notifications@github.com> writes:

> I'm seeing the following denials on RHEL7
>
> [ 19.046773] type=1400 audit(1584424888.832:5): avc: denied { use } for pid=1639 comm="systemd-fstab-g" path="/dev/null" dev="devtmpfs"
> ino=2051 scontext=system_u:system_r:systemd_fstab_generator_t:s0 tcontext=system_u:system_r:kernel_t:s0 tclass=fd permissive=1
> [ 19.048632] type=1400 audit(1584424888.834:6): avc: denied { dac_override } for pid=1639 comm="systemd-fstab-g" capability=1
> scontext=system_u:system_r:systemd_fstab_generator_t:s0 tcontext=system_u:system_r:systemd_fstab_generator_t:s0 tclass=capability permissive=1
> [ 19.048644] type=1400 audit(1584424888.834:7): avc: denied { write } for pid=1639 comm="systemd-fstab-g" name="/" dev="sysfs" ino=1
> scontext=system_u:system_r:systemd_fstab_generator_t:s0 tcontext=system_u:object_r:sysfs_t:s0 tclass=dir permissive=1
> [ 19.048694] type=1400 audit(1584424888.834:8): avc: denied { write } for pid=1639 comm="systemd-fstab-g" name="generator" dev="tmpfs"
> ino=16386 scontext=system_u:system_r:systemd_fstab_generator_t:s0 tcontext=system_u:object_r:systemd_unit_t:s0 tclass=dir permissive=1
> [ 19.048702] type=1400 audit(1584424888.834:9): avc: denied { add_name } for pid=1639 comm="systemd-fstab-g" name="-.mount"
> scontext=system_u:system_r:systemd_fstab_generator_t:s0 tcontext=system_u:object_r:systemd_unit_t:s0 tclass=dir permissive=1
> [ 19.048727] type=1400 audit(1584424888.834:10): avc: denied { create } for pid=1639 comm="systemd-fstab-g" name="-.mount"
> scontext=system_u:system_r:systemd_fstab_generator_t:s0 tcontext=system_u:object_r:systemd_unit_t:s0 tclass=file permissive=1
> [ 19.048741] type=1400 audit(1584424888.834:11): avc: denied { write open } for pid=1639 comm="systemd-fstab-g"
> path="/run/systemd/generator/-.mount" dev="tmpfs" ino=12979 scontext=system_u:system_r:systemd_fstab_generator_t:s0
> tcontext=system_u:object_r:systemd_unit_t:s0 tclass=file permissive=1
>
> It looks like the bulk of the errors are being caused by /run/systemd/generator being labeled systemd_unit_t on my system and this commit is
> expecting that directory to be init_runtime_t. Any idea what is creating the directory /run/systemd/generator?

systemd_unit_t is "appropriate". these are runtime service units, and so
the systemd service access checks are performed on them.  systemd, maybe
tmpfiles using setfscreatecon
and fc specs (possibly also file_contexts.home_dist) are probably
responsible.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

-- 
Key fingerprint = FCD2 3660 5D6B 9D27 7FC6  E0FF DA7E 521F 10F6 4098
https://sks-keyservers.net/pks/lookup?op=get&search=0xDA7E521F10F64098
Dominick Grift

**Comment by dsugar100** [comment]
Thanks that was helpful.  I had an entry in file_contexts.subs_dist that was causing problems.  When I remove that, these problems are going away, but adding problems with the lvm2-activation-generator.  I think that might need to go into a new domain now.

**Pull request merged**